metric-publishers/emf-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/emf/EmfMetricPublisher.java

```diff
@@ -0,0 +1,551 @@
+         * category of metrics uploaded to CloudWatch. Any metrics {@link #publish(MetricCollection)}ed that do not fall under
+         * these configured categories are ignored.
+         *
+         * <p>Note: If there are {@link #dimensions(Collection)} configured that do not fall under these {@code MetricCategory}
+         * values, the dimensions will NOT be ignored. In other words, the metric category configuration only affects which
+         * metrics are uploaded to CloudWatch, not which values can be used for {@code dimensions}.
+         */
+        public Builder metricCategories(Collection<MetricCategory> metricCategories) {
+            this.metricCategories = new ArrayList<>(metricCategories);
+            return this;
+        }
+
+        /**
+         * @see #metricCategories(Collection)
+         */
+        public Builder metricCategories(MetricCategory... metricCategories) {
+            return metricCategories(Arrays.asList(metricCategories));
+        }
+
+        /**
+         * Configure the {@link MetricLevel} that should be uploaded to CloudWatch.
+         *
+         * <p>If this is not specified, {@link MetricLevel#INFO} is used.
+         *
+         * <p>All {@link SdkMetric}s are associated with one {@code MetricLevel}. This setting determines which level of metrics
+         * uploaded to CloudWatch. Any metrics {@link #publish(MetricCollection)}ed that do not fall under these configured
+         * categories are ignored.
+         *
+         * <p>Note: If there are {@link #dimensions(Collection)} configured that do not fall under this {@code MetricLevel}
+         * values, the dimensions will NOT be ignored. In other words, the metric category configuration only affects which
+         * metrics are uploaded to CloudWatch, not which values can be used for {@code dimensions}.
+         */
+        public Builder metricLevel(MetricLevel metricLevel) {
+            this.metricLevel = metricLevel;
+            return this;
+        }
+
+        /**
+         * Configure the LogGroupName key that will be put into the emf log to this publisher. This is a required key for
+         * using the CloudWatch agent to send embedded metric format logs that tells the agent which log group to use.
+         *
+         * <p>If this is not specified, {@code /aws/emf/metrics} will be used.
+         */
+        public Builder logGroupName(String logGroupName) {
+            this.logGroupName = logGroupName;
+            return this;
+        }
+
+
+        /**
+         * Configure the unitTest field to the publisher. It tells the publisher if this publisher is used in unit test.
+         * If so, it will use a fixed timestamp when generating emf string instead of the current time.
+         *
+         * <p>If this is not specified, {@code false} will be used.
+         */
+        public Builder unitTest(boolean unitTest) {
```

Comment:
We should use dependency injection for testing if we need to mock a dependency. eg: https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/GenericS3TransferManager.java#L121